test/e2e/federate.go

```diff
@@ -0,0 +1,160 @@
+			if err != nil {
+				tl.Fatalf("Error initializing dynamic client: %v", err)
+			}
+		}
+	})
+
+	// Use one cluster scoped and one namespaced type to test complete flow
+	toTest := []string{"clusterroles.rbac.authorization.k8s.io", "configmaps"}
+	for _, testKey := range toTest {
+		typeConfigName := testKey
+		fixture := typeConfigFixtures[testKey]
+		It(fmt.Sprintf("for %q should create an equivalant federated resource in federation", typeConfigName), func() {
+			typeConfig := &fedv1a1.FederatedTypeConfig{}
+			err := client.Get(context.Background(), typeConfig, f.FederationSystemNamespace(), typeConfigName)
+			if err != nil {
+				tl.Fatalf("Error retrieving federatedtypeconfig %q: %v", typeConfigName, err)
+			}
+
+			if framework.TestContext.LimitedScope && !typeConfig.GetNamespaced() {
+				framework.Skipf("Federation of cluster-scoped type %s is not supported by a namespaced control plane.", typeConfigName)
+			}
+
+			kind := typeConfig.GetTarget().Kind
+			targetAPIResource := typeConfig.GetTarget()
+			targetObject, err := common.NewTestTargetObject(typeConfig, f.TestNamespaceName(), fixture)
+			if err != nil {
+				tl.Fatalf("Error creating test object: %v", err)
+			}
+
+			createdTargetObject, err := common.CreateResource(kubeConfig, targetAPIResource, targetObject)
+			if err != nil {
+				tl.Fatalf("Error creating resource: %v", err)
+			}
+
+			typeName := util.QualifiedName{
+				Name:      typeConfig.GetObjectMeta().Name,
+				Namespace: typeConfig.GetObjectMeta().Namespace,
+			}
+			testResourceName := util.NewQualifiedName(createdTargetObject)
+
+			defer deleteResources(f, tl, typeConfig, testResourceName)
+
+			tl.Logf("Federating %s %q", kind, testResourceName)
+			federatedObject, err := federate.FederateResource(kubeConfig, typeName, testResourceName, false)
+			if err != nil {
+				tl.Fatalf("Error federating %s %q: %v", kind, testResourceName, err)
+			}
+
+			validateTemplateEquality(tl, federatedObject, createdTargetObject, typeConfig.GetFederatedType().Kind)
+		})
+	}
+})
+
+func validateTemplateEquality(tl common.TestLogger, fedObj, targetObj *unstructured.Unstructured, fedKind string) {
+	qualifiedName := util.NewQualifiedName(fedObj)
+	templateMap, ok, err := unstructured.NestedFieldCopy(fedObj.Object, util.SpecField, util.TemplateField)
+	if err != nil || !ok {
+		tl.Fatalf("Error retrieving template from %s %q", fedKind, qualifiedName)
+	}
+
+	expectedObj := &unstructured.Unstructured{}
+	expectedObj.Object = templateMap.(map[string]interface{})
+	federate.RemoveUnwantedFields(expectedObj)
+	federate.RemoveUnwantedFields(targetObj)
+
+	if !reflect.DeepEqual(expectedObj, targetObj) {
+		tl.Fatal("Federated object template and target object don't match for %s %q", fedKind, qualifiedName)
+	}
+}
+
+func deleteResources(f framework.FederationFramework, tl common.TestLogger, typeConfig typeconfig.Interface, testResourceName util.QualifiedName) {
+	fedAPIResource := typeConfig.GetFederatedType()
```

Comment:
(No action required) Why is it necessary to attempt deletion of the federated resource if it is namespaced?